src/algebra/commute.lean

```diff
@@ -0,0 +1,267 @@
+/-
+Copyright (c) 2019 Neil Strickland. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Neil Strickland
+
+Commuting pairs of elements in not-necessarily-commutative monoids,
+groups and (semi) rings.  Centralizers as submonoids, subgroups or
+subrings.
+
+Usage example: suppose that a, b and c are elements of a ring, and
+that `hb : commute a b` and `hc : commute a c`.  Then
+`hb.pow_left 5` proves `commute (a ^ 5) b` and
+`(hb.pow 2).add (hb.mul hc)` proves `commute a (b ^ 2 + b * c)`.
+Lean does not immediately recognise these terms as equations,
+so for rewriting we need syntax like `rw [(hb.pow_left 5).eq]`
+rather than just `rw [hb.pow_left 5]`.
+-/
+
+import algebra.group_power
+  group_theory.submonoid group_theory.subgroup
+  ring_theory.subring
+
+section semigroup
+
+variables {S : Type*} [semigroup S]
+
+@[reducible] def commute (a b : S) : Prop := a * b = b * a
```

Comment:
Does anything break if this is not `@[reducible]`?

Reply:
I'll look into this.

Reply:
It seems that the choice is:
* either `commute` is `@[reducible]`, and we can use `h : commute a b` in `rw [h]` directly;
* or `commute` is not `@[reducible]`, and we need to write `rw [h.eq]`.

Reply:
I see. I saw the documentation that says you need to use `.eq`, and so I assumed `reducible` didn't actually achieve that, and then I wondered what else it could be doing.

Reply:
Anyways, that means we should either make it not `[reducible]`, or delete the part about using `.eq` from the module documentation, and document here that `[reducible]` is so that you can use it with `rw`. (Same with `simp`, btw?)

Reply:
For now, I removed `@[reducible]`.